### PR TITLE
Custom validations

### PR DESCRIPTION
**Pull request by rodzyn**
Small refactoring + basic custom validators. Starting point for more improvements like more OO implementation of existing (predefined) list of validations.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,10 @@
 # Lotus::Validations
 Validations mixin for Ruby objects
 
+## v0.2.5
+### Added
+- [Marcin Olichwirowicz] Custom Validators
+
```
Comment (AlfonsoUceda) [inline CHANGELOG.md]: This should be added when @jodosha bump the version

File: Gemfile [hunk 2]
```diff
@@ -10,3 +10,7 @@ gem 'lotus-utils', require: false, github: 'lotus/utils', branch: '0.3.x'
 gem 'lotus-model', require: false, github: 'lotus/model', branch: '0.2.x'
 gem 'simplecov',   require: false
 gem 'coveralls',   require: false
+
+group :development do
+  gem 'pry'
```
Comment (AlfonsoUceda) [inline Gemfile]: pry should be removed here

File: README.md [hunk 3]
```diff
@@ -356,6 +356,39 @@ signup.valid? # => false
...
+  include Lotus::Validations::Validator
+
+  def valid?(value)
+    !value.empty?
+  end
+end
+
+```
+
+And utilise it using `with` option
```
Comment (AlfonsoUceda) [inline README.md]: s/utilise/utilize

File: README.md [hunk 4]
```diff
@@ -356,6 +356,39 @@ signup.valid? # => false
...
+
+```ruby
+require 'lotus/validations'
+
+class User
+  attribute :name, with: CustomPresenceValidator
+end
+
+user = User.new(name: '')
+user.valid? # => falase
```
Comment (AlfonsoUceda) [inline README.md]: s/falase/false

File: lib/lotus/validations/validation_set.rb [hunk 5]
```diff
@@ -53,7 +54,18 @@ def names
         @validations.keys
       end
 
+      def validate(attributes, errors)
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validation_set.rb]: Please document this method

File: CHANGELOG.md (hunk 1 above)
Comment (jodosha) [inline CHANGELOG.md]: :+1: to @AlfonsoUceda I'll take care of this. :wink:

**Comment by jodosha** [comment]
@rodzyn Thank you for this PR.

I'm wondering why the validator has a different approach to validations. That `#valid?` method that exposes reminds me a FP style. I don't mind FP at all, but there are some aspects to consider.

`Lotus::Validations` works like this `Signup.new(attributes).valid?`, while `Lotus::Validations::Validator` does `PresenceValidator.new.valid?(value)`. This is an inconsistent approach.

In other words, we should probably do `PresenceValidator.new(value).valid?` instead.
The reason isn't just consistency with the rest of the library, but because this FP style makes hard to compose a complex custom validator.

In Italy we have a Social Security Number called [Codice Fiscale](http://en.wikipedia.org/wiki/Italian_fiscal_code_card). It's an unique identifier for citizens.
It has a complex set of validations rules:

  * Length of 16
  * Format: `FFFLLL00M11C222X`
  * The `11` placeholder above is the exact day of birth for men, and you add `40` for women.
  * Etc..

```ruby
# FP style.
class CFValidator
  def valid?(value)
    valid_length?(value) &&
      valid_format?(value) &&
      valid_birth_date?(value)
  end
end
```

vs

```ruby
# OOP style
class CFValidator
  def initialize(value)
    @value = value
  end

  def valid?
    valid_length? &&
      valid_format? &&
      valid_birth_date?
  end
end
```

In the end, I'm wondering how a developer can return error messages from it. Eg. "Birth date is not valid".

Am I missing something? If so, please make it more clear by adding additional tests. Thank you very much! :hamster:

**Comment by rodzyn** [comment]
@AlfonsoUceda Nice catches!
@jodosha Thanks for detailed response. 

Regarding inconsistency, I was focused on this level now:

```ruby

class PresenceValidator
  include Lotus::Validations::Validator

  def valid?
    !@value.empty?
  end
end

class Signup
  include Lotus::Validations

  attribute :name, with: PresenceValidator
end
```


```ruby
Signup.new(name: '').valid? #false
```

And old version:

```ruby
class Signup
  include Lotus::Validations

  attribute :name, presence: true
end
```

```ruby
Signup.new(name: '').valid?
```

I was thinking about OOP vs FP comparison and I've slightly changed implementation, so now you can easily do:

```ruby
PresenceValidator.new('').valid?
```

And we ended up with good consistency I guess :)

Part about error handling is the thing what I was planning to implement(gradually) in the next PR. It needs some changes in current validation process. This PR is the first step :)

**Comment by jodosha** [comment]
@rodzyn Hey, thank you for the quick refactoring.

Can you please keep continue working on this PR for adding validations messages?

I would love to keep master clean as now. If we need to do a quickfix release, then this half implemented feature would be shipped with the stable code. Does it makes sense?

**Comment by rodzyn** [comment]
@jodosha Sounds good :+1:

**Comment by rodzyn** [comment]
Hi again, quick update about the progress.
To validate the idea of custom validators and simplify code base I've implemented
all of current validators using my custom validator solution.
Example:

```ruby
module Lotus
  module Validations
    module Validator
      class Acceptance
        include Lotus::Validations::Validator

        def valid?
          Lotus::Utils::Kernel.Boolean(@value)
        end

        def validator_name
          :acceptance
        end
      end
    end
  end
end
```
There are some lacks in documentation and many things could be simplified a bit more - I'm currently pushing it as a draft for you to take a look and maybe share some comments about the idea.

Regarding:

> In the end, I'm wondering how a developer can return error messages from it. Eg. "Birth date is not valid"

&

> Can you please keep continue working on this PR for adding validations messages?

I went through documentation and I don't see any existing solution to provide such
a message so I guess it needs to be implemented from scratch?

During changes made above, I found quite hard to deal with two types of validators: confirmation and nested. E.g.: I had to implement kind of fake validator for "type validation" (coercion) which is also passed to validators set.
I also found a lot of places with some bad hacks around confirmation and nested stuff. 

All these things make me think about re-implementation of the core part like DSL, etc.
I'm also wondering if it's not better to use Virtus for coercion, then here just make validation gem. I would also get rid of compatibility with Lotus::Entity for now and found better solution like some proper interfaces or separate gem.

What are your thoughts about that? /cc @jodosha

File: test/custom_validation_test.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    it "can mimic presence validator" do
+      validator = CustomPresenceValidation.new(name: '')
+      validator.valid?.must_equal false
+      validator.name = "X"
+      validator.valid?.must_equal true
+    end
+  end
+
+  describe '#validate keyword' do
+    it "can mimic presence validator" do
```
Comment (jodosha) [inline test/custom_validation_test.rb]: Is this `describe` block a duplication?

**Comment by jodosha** [comment]
> I went through documentation and I don't see any existing solution to provide such
> a message so I guess it needs to be implemented from scratch?

Yes, those exist only as error primitive (`Error`). We haven't implemented yet the presentational aspect of those messages.

**Comment by jodosha** [comment]
> All these things make me think about re-implementation of the core part like DSL, etc.

Yes, we're experiencing an high churn rate. I'm worried about this, but we're stretching the initial code of this gem. I'm positive that we will find internal stability over time. Good testing coverage is crucial for this.

> I'm also wondering if it's not better to use Virtus for coercion, then here just make validation gem.

Virtus is a great tool, but it would introduce a lot of runtime dependencies for Lotus::Validations:

  * virtus
  * axiom-types
  * coercible
  * descendants_tracker
  * thread_safe
  * ice_nine
  * equalizer

This seems to be too much for a single feature. Also, Lotus::Controller depends on this gem to power the optional feature of params validation. It will force developers to install all those gems for something that may or may not use.

> I would also get rid of compatibility with Lotus::Entity for now and found better solution like some proper interfaces or separate gem.

Why do you think that the integration with `Lotus::Entity` should be dropped? Personally, I don't think that validations should be handled there. However, interoperability between Lotus gems is something that we strive for. Please remember that modularity is at the core.

Thank you for your thoughtful considerations of the scope of Lotus::Validations, but those concerns listed above have a greater weight than custom validations, which are a nice to have feature.

File: test/custom_validation_test.rb (hunk 6 above)
Comment (rodzyn) [inline test/custom_validation_test.rb]: One test is for `CustomPresenceValidation` and another for `CustomAttributesValidation`. I'll do a general cleanup & maybe some more testing soon!

**Comment by plukevdh** [comment]
I'd love to see this shipped first before working out how error messaging, as that seems a separate concern. I'd love to take it on, but if this is going to be included at some point, any work I would do could change radically.

**Comment by jodosha** [comment]
@rodzyn @plukevdh Even if the CI is green there are some issues. For instance, we dropped attribute name for `Error`. That is important when you iterate errors.

```ruby
class Signup
  include Lotus::Validations

  attribute :first_name, presence: true
  attribute :last_name, presence: true
end

validator = Signup.new
validator.valid? # => false

# All the errors here have the same values, without any chance to link them with the failing attribute.
validator.errors.each do |error|
  puts error.validation # => :presence
  puts error.expected # => true
  puts error.actual # => nil
end
```

File: Gemfile [hunk 7]
```diff
@@ -10,3 +10,4 @@ gem 'lotus-utils', require: false, github: 'lotus/utils', branch: '0.3.x'
 gem 'lotus-model', require: false, github: 'lotus/model', branch: '0.2.x'
 gem 'simplecov',   require: false
 gem 'coveralls',   require: false
+gem 'pry'
```
Comment (jeremyf) [inline Gemfile]: I don't believe this should be added.

**Comment by joneslee85** [comment]
@rodzyn ping, can you please rebase?

**Comment by rodzyn** [comment]
@joneslee85 yeah, sure. I hope to find some time this evening/night (european time zone)

**Comment by rodzyn** [comment]
Rebased

File: Gemfile [hunk 8]
```diff
@@ -10,3 +10,4 @@ gem 'lotus-utils', require: false, github: 'lotus/utils', branch: '0.4.x'
 gem 'lotus-model', require: false, github: 'lotus/model', branch: '0.3.x'
 gem 'simplecov',   require: false
 gem 'coveralls',   require: false
+gem 'pry'
```
Comment (joneslee85) [inline Gemfile]: This can be removed once done

**Comment by joneslee85** [comment]
@rodzyn cheers, I will review it shortly

File: Gemfile (hunk 8 above)
Comment (AlfonsoUceda) [inline Gemfile]: remove this dependency please

File: lib/lotus/validations/attribute.rb [hunk 9]
```diff
@@ -199,29 +203,27 @@ def size
...
+          @errors = Lotus::Validations::Validator::Nested.call(@value)
+
           true
         end
       end
 
-      # @since 0.1.0
+      # Validates custom Lotus Validator objects
+      #
+      # @since 0.2.5
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/attribute.rb]: replace with x.x.x

File: lib/lotus/validations/error.rb [hunk 10]
```diff
@@ -63,21 +63,16 @@ class Error
...
       #
-      # @param attribute_name [Symbol] the name of the attribute
       # @param validation [Symbol] the name of the validation
       # @param expected [Object] the expected value
       # @param actual [Object] the actual value
-      # @param namespace [String] the optional namespace
       #
       # @since 0.1.0
       # @api private
-      def initialize(attribute_name, validation, expected, actual, namespace = nil)
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/error.rb]: why namespace is removed?

File: lib/lotus/validations/error.rb [hunk 11]
```diff
@@ -63,21 +63,16 @@ class Error
...
-        @attribute_name = attribute_name.to_s
+      def initialize(validation, expected, actual, namespace = nil)
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/error.rb]: remove namespace parameter

File: lib/lotus/validations/errors.rb [hunk 12]
```diff
@@ -68,20 +70,7 @@ def clear
...
-      # Iterate thru the errors, yields the given block and collect the
-      # returning value.
-      #
-      # @param blk [Proc] the given block
-      # @yield [error] a Lotus::Validations::Error
-      #
-      # @see Lotus::Validations::Error
-      #
-      # @since 0.1.0
-      def map(&blk)
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/errors.rb]: why did you remove `map`

File: lib/lotus/validations/validation_set.rb [hunk 13]
```diff
@@ -53,7 +54,25 @@ def names
         @validations.keys
       end
 
+      # For each attribute it creates new Attribute object and runs all
+      # associated validations
+      #
+      # @since 0.2.5
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validation_set.rb]: replace with x.x.x

File: lib/lotus/validations/validator/presence.rb [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+  module Validations
+    module Validator
+      class Presence
+        include Lotus::Validations::Validator
+
+        BLANK_STRING_MATCHER = /\A[[:space:]]*\z/.freeze
+
+        # Checks if the value is "blank".
+        #
+        # @since 0.2.2
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/presence.rb]: replace all 0.2.2 with x.x.x

File: lib/lotus/validations/validator/presence.rb [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+          @value.respond_to?(:match) && @value.match(BLANK_STRING_MATCHER)
+        end
+
+        # @since 0.2.2
+        # @api private
+        def _empty_value?
+          @value.respond_to?(:empty?) && @value.empty?
+        end
+
+
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/presence.rb]: remove these lines

File: lib/lotus/validations/validator/acceptance.rb [hunk 16]
```diff
@@ -0,0 +1,17 @@
+module Lotus
+  module Validations
+    module Validator
+      class Acceptance
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/acceptance.rb]: comment this class

File: lib/lotus/validations/validator/confirmation.rb [hunk 17]
```diff
@@ -0,0 +1,17 @@
+module Lotus
+  module Validations
+    module Validator
+      class Confirmation
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/confirmation.rb]: comment this class

File: lib/lotus/validations/validator/exclusion.rb [hunk 18]
```diff
@@ -0,0 +1,19 @@
+module Lotus
+  module Validations
+    module Validator
+      class Exclusion
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/exclusion.rb]: comment this class

File: lib/lotus/validations/validator/format.rb [hunk 19]
```diff
@@ -0,0 +1,19 @@
+module Lotus
+  module Validations
+    module Validator
+      class Format
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/format.rb]: comment this class

File: lib/lotus/validations/validator/inclusion.rb [hunk 20]
```diff
@@ -0,0 +1,19 @@
+module Lotus
+  module Validations
+    module Validator
+      class Inclusion
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/inclusion.rb]: comment this class

File: lib/lotus/validations/validator/nested.rb [hunk 21]
```diff
@@ -0,0 +1,21 @@
+module Lotus
+  module Validations
+    module Validator
+      class Nested
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/nested.rb]: comment this class

File: lib/lotus/validations/validator/presence.rb [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+        # Checks if the value is "blank".
+        #
+        # @since 0.2.2
+        # @api private
+        def valid?
+          !(_nil_value? || _blank_string? || _empty_value?)
+        end
+
+        def validator_name
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/presence.rb]: comment this methods

File: lib/lotus/validations/validator/size.rb [hunk 23]
```diff
@@ -0,0 +1,26 @@
+module Lotus
+  module Validations
+    module Validator
+      class Size
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/size.rb]: comment this class

File: lib/lotus/validations/validator/type.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+module Lotus
+  module Validations
+    module Validator
+      class Type
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/type.rb]: comment this class

File: lib/lotus/validations/validator/with.rb [hunk 25]
```diff
@@ -0,0 +1,17 @@
+module Lotus
+  module Validations
+    module Validator
+      class With
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/validator/with.rb]: comment this class

File: test/exclusion_validation_test.rb [hunk 26]
```diff
@@ -4,7 +4,7 @@
   describe 'exclusion' do
     it "is valid if attribute value is escluded in specified enumerable" do
       validator = ExclusionValidatorTest.new(job: 'Programmer', age: '17', state: 'RM', sport: 'Snorkeling', code: 'j')
-
+
```
Comment (AlfonsoUceda) [inline test/exclusion_validation_test.rb]: remove this line

File: test/initialize_test.rb [hunk 27]
```diff
@@ -44,9 +44,10 @@
         'password_confirmation' => '123',
         'unknown'               => 'blah'
       )
-
+
```
Comment (AlfonsoUceda) [inline test/initialize_test.rb]: remove this line or indentation

File: test/size_validation_test.rb [hunk 28]
```diff
@@ -68,13 +68,13 @@
         File.new(__FILE__),
         Pathname.new(__FILE__)
       ]
-
+
```
Comment (AlfonsoUceda) [inline test/size_validation_test.rb]: remove this line or indentation

File: test/test_helper.rb [hunk 29]
```diff
@@ -27,3 +27,4 @@ def defined_validation?(name)
 end
 
 require 'fixtures'
+require 'pry'
```
Comment (AlfonsoUceda) [inline test/test_helper.rb]: remove this line

**Comment by AlfonsoUceda** [comment]
@rodzyn can you take care of comment? thanks ;) +1 for me ;)

File: lib/lotus/validations/errors.rb (hunk 12 above)
Comment (rodzyn) [inline lib/lotus/validations/errors.rb]: Since we have it as `Enumerable` we don't need to implement `map`, `each` is enough.

**Comment by rodzyn** [comment]
I think the example from @jodosha is not align with current state here. It looks rather like this, so we do have information about connection between attribute and errors:

```ruby
class Signup
  include Lotus::Validations

  attribute :first_name, exclusion: 'z', inclusion: 'x'
  attribute :last_name, presence: true
end

validator = Signup.new(first_name: 'z')
validator.valid? #=> false

validator.errors

=> #<Lotus::Validations::Errors:0x007fbb6485daf0
 @errors=
  {:first_name=>
    [#<Lotus::Validations::Error:0x007fbb6485d8c0 @actual="z", @expected="z", @validation=:exclusion>,
     #<Lotus::Validations::Error:0x007fbb6485d668 @actual="z", @expected="x", @validation=:inclusion>],
   :last_name=>[#<Lotus::Validations::Error:0x007fbb6485cf88 @actual=nil, @expected=true, @validation=:presence>]}>

validator.errors.each { |attr, errors| p "#{attr} - #{errors.map(&:validation)}"}

#"first_name - [:exclusion, :inclusion]"
#"last_name - [:presence]"

validator.errors.for(:first_name)
=> [#<Lotus::Validations::Error:0x007fbb6485d8c0 @actual="z", @expected="z", @validation=:exclusion>,
 #<Lotus::Validations::Error:0x007fbb6485d668 @actual="z", @expected="x", @validation=:inclusion>]

validator.errors.for(:last_name)
=> [#<Lotus::Validations::Error:0x007fbb6485cf88 @actual=nil, @expected=true, @validation=:presence>]

```

**Comment by rodzyn** [comment]
btw. there is some odd travis behaviour for one of the builds (the one is failing)

**Comment by joneslee85** [comment]
@rodzyn can you please tweet to Travis and see if they could help? After green build, I'll review it one more time

**Comment by rodzyn** [comment]
@joneslee85 I'm trying to reach them on irc

**Comment by rodzyn** [comment]
@joneslee85 Finally, I found a solution by myself - cleaning bundler cache in Travis

File: lib/lotus/validations/attribute.rb [hunk 30]
```diff
@@ -199,29 +199,27 @@ def size
...
       def nested
         _validate(__method__) do |validator|
-          errors = value.validate
-          errors.each do |error|
-            new_error = Error.new(error.attribute, error.validation, error.expected, error.actual, @name)
-            @errors.add new_error.attribute, new_error
-          end
+
+          @errors = Lotus::Validations::Validator::Nested.call(@value)
+
```
Comment (joneslee85) [inline lib/lotus/validations/attribute.rb]: _nitpick_ remove this empty line

File: lib/lotus/validations/errors.rb (hunk 12 above)
Comment (joneslee85) [inline lib/lotus/validations/errors.rb]: :+1:

File: test/attribute_test.rb [hunk 31]
```diff
@@ -16,10 +16,10 @@
...
 
       errors = validator.errors.for(:email)
-      errors.must_equal [
-        Lotus::Validations::Error.new(:email, :format, /@/, 'test'),
-        Lotus::Validations::Error.new(:email, :confirmation, true, 'test')
-      ]
+
+      errors.must_include Lotus::Validations::Error.new(:format, /@/, 'test')
+      errors.must_include Lotus::Validations::Error.new(:confirmation, true, 'test')
+
```
Comment (joneslee85) [inline test/attribute_test.rb]: nitpick, remove this empty line

File: test/custom_validation_test.rb (hunk 6 above)
Comment (joneslee85) [inline test/custom_validation_test.rb]: can you please clean this up?

**Comment by joneslee85** [comment]
Except above comments, I love what I see, much thanks, :+1:

File: lib/lotus/validations/error.rb [hunk 32]
```diff
@@ -63,21 +60,16 @@ class Error
...
       # @param validation [Symbol] the name of the validation
       # @param expected [Object] the expected value
       # @param actual [Object] the actual value
-      # @param namespace [String] the optional namespace
       #
       # @since 0.1.0
       # @api private
-      def initialize(attribute_name, validation, expected, actual, namespace = nil)
-        @attribute_name = attribute_name.to_s
+      def initialize(validation, expected, actual)
```
Comment (jodosha) [inline lib/lotus/validations/error.rb]: @rodzyn Why the attribute name is removed from here?

Comment (jodosha) [inline lib/lotus/validations/error.rb]: @rodzyn More on this. In your [comment above](https://github.com/lotus/validations/pull/49#issuecomment-106631885) I see a snippet of code:

```ruby
validator.errors

=> #<Lotus::Validations::Errors:0x007fbb6485daf0
 @errors=
  {:first_name=>
    [#<Lotus::Validations::Error:0x007fbb6485d8c0 @actual="z", @expected="z", @validation=:exclusion>,
     #<Lotus::Validations::Error:0x007fbb6485d668 @actual="z", @expected="x", @validation=:inclusion>],
   :last_name=>[#<Lotus::Validations::Error:0x007fbb6485cf88 @actual=nil, @expected=true, @validation=:presence>]}>
```

When we do `errors.each` we yield a flatten array of `Error` instances. We loose that data structure (`Hash`), where the attribute name is the key for an array of errors (`:first_name => [...]`).

By removing attribute name, the `Error` instances that we yield, **they loose the reference with the attribute**. Look at the following error, how you can tell which is the attribute that caused it?

```ruby
#<Lotus::Validations::Error:0x007fbb6485d8c0 @actual="z", @expected="z", @validation=:exclusion>
```

Can you please restore it? Thank you.

Comment (rodzyn) [inline lib/lotus/validations/error.rb]: Actually you don't need to have flattened collection of errors, you can use:
```ruby
validator.errors.each { |attr, errors| }
```
And in `attr` you have a reference with the attribute. That's why I removed it from the error object - because of redundant information. Probably I'm missing some downsides of my approach. What do you think about this? I'd love to hear some valid points against it before restoring it.

Comment (jodosha) [inline lib/lotus/validations/error.rb]: @rodzyn Consider the following code, it uses v0.3.2

```ruby
 ~:ruby-2.2.2 ➜ pry
[1] pry(main)> require 'lotus/validations'
=> true
[2] pry(main)> class Signup
[2] pry(main)*   include Lotus::Validations
[2] pry(main)*   attribute :user do
[2] pry(main)*     attribute :name, presence: true
[2] pry(main)*     attribute :address do
[2] pry(main)*       attribute :city, presence: true
[2] pry(main)*     end
[2] pry(main)*   end
[2] pry(main)*   attribute :terms, acceptance: true
[2] pry(main)* end
=> {:acceptance=>true}
[3] pry(main)> signup = Signup.new
=> #<Signup:0x007fa4c4c4bb28 @attributes=#<Lotus::Utils::Attributes:0x007fa4c4c4bad8 @attributes={}>>
[4] pry(main)> signup.valid?
=> false
[5] pry(main)> signup.errors
=> #<Lotus::Validations::Errors:0x007fa4c4c23998
 @errors=
  {"user.name"=>
    [#<Lotus::Validations::Error:0x007fa4c4c22c50 @actual=nil, @attribute="user.name", @attribute_name="name", @expected=true, @namespace=:user, @validation=:presence>],
   "user.address.city"=>
    [#<Lotus::Validations::Error:0x007fa4c4c22ae8
      @actual=nil,
      @attribute="user.address.city",
      @attribute_name="address.city",
      @expected=true,
      @namespace=:user,
      @validation=:presence>],
   :terms=>[#<Lotus::Validations::Error:0x007fa4c4c228e0 @actual=nil, @attribute="terms", @attribute_name="terms", @expected=true, @namespace=nil, @validation=:acceptance>]}>
[6] pry(main)> signup.errors.each do |error|
[6] pry(main)*   puts error.attribute
[6] pry(main)*   puts error.validation
[6] pry(main)*   puts "\n\n\n"
[6] pry(main)* end
user.name
presence



user.address.city
presence



terms
acceptance



=> [#<Lotus::Validations::Error:0x007fa4c4c22c50 @actual=nil, @attribute="user.name", @attribute_name="name", @expected=true, @namespace=:user, @validation=:presence>,
 #<Lotus::Validations::Error:0x007fa4c4c22ae8
  @actual=nil,
  @attribute="user.address.city",
  @attribute_name="address.city",
  @expected=true,
  @namespace=:user,
  @validation=:presence>,
 #<Lotus::Validations::Error:0x007fa4c4c228e0 @actual=nil, @attribute="terms", @attribute_name="terms", @expected=true, @namespace=nil, @validation=:acceptance>]
[7] pry(main)>
```

The advantage of having flatten errors is the use only one level of `each`.

With your suggestion you need to know exactly how many levels of depth and construct nested loops. In other words, the way we should iterate for errors, will change, if we change the nested structure of the validations. Which is an annoying maintenance issue.

```ruby
irb(main):001:0> class Signup
irb(main):002:1> include Lotus::Validations
irb(main):003:1> attribute :user do
irb(main):004:2* attribute :name, presence: true
irb(main):005:2> attribute :address do
irb(main):006:3* attribute :city, presence: true
irb(main):007:3> end
irb(main):008:2> end
irb(main):009:1> attribute :terms, acceptance: true
irb(main):010:1> end
=> {:acceptance=>true}
irb(main):011:0> signup = Signup.new
=> #<Signup:0x007fee19d9cdf8 @attributes=#<Lotus::Utils::Attributes:0x007fee19d9cd80 @attributes={}>>
irb(main):012:0> signup.valid?
=> false
irb(main):013:0> signup.errors
=> #<Lotus::Validations::Errors:0x007fee19d95c38 @errors={"user.name"=>[#<Lotus::Validations::Error:0x007fee19d958a0 @validation=:presence, @expected=true, @actual=nil>], "user.address.city"=>[#<Lotus::Validations::Error:0x007fee19d955a8 @validation=:presence, @expected=true, @actual=nil>], :terms=>[#<Lotus::Validations::Error:0x007fee19d94f90 @validation=:acceptance, @expected=true, @actual=nil>]}>
irb(main):014:0> signup.errors.each do |attr_name, errors|
irb(main):015:1* puts attr_name
irb(main):016:1> errors.each do |error|
irb(main):017:2* puts error.validation
irb(main):018:2> end
irb(main):019:1> puts "\n\n\n"
irb(main):020:1> end
user.name
presence



user.address.city
presence



terms
acceptance



=> {"user.name"=>[#<Lotus::Validations::Error:0x007fee19d958a0 @validation=:presence, @expected=true, @actual=nil>], "user.address.city"=>[#<Lotus::Validations::Error:0x007fee19d955a8 @validation=:presence, @expected=true, @actual=nil>], :terms=>[#<Lotus::Validations::Error:0x007fee19d94f90 @validation=:acceptance, @expected=true, @actual=nil>]}
```

Imagine we want to introduce an `ErrorPresenter` (which is something we are considering). Its constructor should accept an `Error` and an attribute name. It looks like a Data Clump. Error is a type designed to remove this problem and have only one object to hold all the informations, without any external reference (the attr name).

Comment (rodzyn) [inline lib/lotus/validations/error.rb]: Ok, thanks for a great examples, the idea behind it is now much more clear for me! I'm gonna work now on getting back to this structure.

**Comment by jodosha** [comment]
@rodzyn Closing this, as it's too much outdated. Thanks for your help :green_heart: :blue_heart: :yellow_heart:

**Comment by leandrocp** [comment]
Amazing PR :+1: 

@rodzyn can I help somehow ?
